File: apps/waiter/src/services/order.service.ts
import api from './api';

interface CreateOrderPayload {
  type: string;
  tableId?: string;
  guestCount?: number;
  items: { productId: string; quantity: number }[];
}

interface OrdersResponse {
  data: {
    id: string;
    orderNumber: string;
    status: string;
    total: number;
    createdAt: string;
    items: {
      id: string;
      quantity: number;
      price: number;
      product?: { id: string; name: string; price: number };
    }[];
  }[];
}

export const orderService = {
  getAll: async (): Promise<OrdersResponse> => {
    const { data: response } = await api.get('/orders');
    // API returns { success, data: [...orders] }
    return { data: response.data || [] };
  },

  getByTableId: async (tableId: string): Promise<OrdersResponse> => {
    const { data: response } = await api.get(`/orders?tableId=${tableId}&limit=50`);
    const all = response.data || [];
    // Only show active orders (not completed/cancelled)
    const active = all.filter((o: any) =>
      !['COMPLETED', 'CANCELLED'].includes(o.status)
    );
    return { data: active };
  },

  create: async (payload: CreateOrderPayload): Promise<{ data: { id: string } }> => {
    const { data: response } = await api.post('/orders', payload);
    return { data: response.data };
  },

  addItems: async (orderId: string, items: { productId: string; quantity: number }[]): Promise<void> => {
    await api.post(`/orders/${orderId}/items`, { items });
  },

  updateStatus: async (orderId: string, status: string): Promise<void> => {
    await api.patch(`/orders/${orderId}/status`, { status });
  },

  updateItemQuantity: async (orderId: string, itemId: string, quantity: number): Promise<void> => {
    await api.patch(`/orders/${orderId}/items/${itemId}`, { quantity });
  },

  printReceipt: async (orderId: string): Promise<void> => {
    await api.post(`/printer/print/receipt/${orderId}`);
  },
};

File: apps/waiter_flutter/lib/services/order_service.dart
import '../core/api.dart';
import '../models/order.dart';

class OrderService {
  Future<List<Order>> getAll({String? status}) async {
    final Map<String, dynamic> queryParams = {};
    if (status != null) queryParams['status'] = status;

    final response = await dio.get('/orders', queryParameters: queryParams.isNotEmpty ? queryParams : null);
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final List<dynamic> list = body['data'] as List<dynamic>? ??
        body['orders'] as List<dynamic>? ??
        (data is List ? data : []);
    return list.map((e) => Order.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<Order>> getByTableId(String tableId) async {
    final response = await dio.get('/orders', queryParameters: {'tableId': tableId});
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final List<dynamic> list = body['data'] as List<dynamic>? ??
        body['orders'] as List<dynamic>? ??
        (data is List ? data : []);
    return list.map((e) => Order.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Order> create({
    required String tableId,
    required int guestCount,
    required List<Map<String, dynamic>> items,
  }) async {
    final response = await dio.post('/orders', data: {
      'type': 'DINE_IN',
      'tableId': tableId,
      'guestCount': guestCount,
      'items': items,
    });
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final orderData = body['data'] ?? body;
    return Order.fromJson(orderData as Map<String, dynamic>);
  }

  Future<Order> addItems(String orderId, List<Map<String, dynamic>> items) async {
    final response = await dio.post('/orders/$orderId/items', data: {'items': items});
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final orderData = body['data'] ?? body;
    return Order.fromJson(orderData as Map<String, dynamic>);
  }

  Future<Order> updateItemQuantity(String orderId, String itemId, int quantity) async {
    final response = await dio.patch('/orders/$orderId/items/$itemId', data: {'quantity': quantity});
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final orderData = body['data'] ?? body;
    return Order.fromJson(orderData as Map<String, dynamic>);
  }

  Future<Order> updateStatus(String orderId, String status) async {
    final response = await dio.patch('/orders/$orderId/status', data: {'status': status});
    final data = response.data;
    final Map<String, dynamic> body = data is Map<String, dynamic> ? data : {};
    final orderData = body['data'] ?? body;
    return Order.fromJson(orderData as Map<String, dynamic>);
  }

  Future<void> printReceipt(String orderId) async {
    await dio.post('/printer/print/receipt/$orderId');
  }

  Future<void> closeTable(String tableId) async {
    await dio.post('/tables/$tableId/close');
  }
}
